bindings/js/api.js

```diff
@@ -0,0 +1,283 @@
+      - 'args', the list of the WebAssembly arguments expected by the function
+        - 'name', the name of the argument which will be shown in the JS Doc
+        - 'kind', either `input` or `output` of the function
+        - 'type', either 'int' or 'boolean' or 'buffer'
+        - 'size', either an integer of a string which is the 'name' of another
+          argument of type 'int'
+        - 'interface_index', for all `input` that should appear in JS, position
+          inside the argument list of the JS function
+        - 'tests', a list of values for this arguments, each value corresponding
+          to a different test case
+      - 'return', the return type of the WebAssembly function
+  */
+
+  const CheckIfByteArray = function(candidate, length, name) {
+    if (!(typeof(candidate) === "object") ||
+      !(candidate.length === length) ||
+      !(candidate.constructor === Uint8Array)
+    ) {
+      throw new Error(
+        "name: Please ensure the argument " + candidate + " is a " + length + "-bytes Uint8Array."
+      );
+    }
+  };
+
+  const copy_array_to_stack = function(array) {
+    let pointer = loader.reserve(Module.Kremlin.mem, array.length);
+    (new Uint8Array(Module.Kremlin.mem.buffer)).set(array, pointer);
+    return pointer;
+  };
+
+  const read_memory = function(ptr, len) {
+    var result = new ArrayBuffer(len);
+    (new Uint8Array(result).set(new Uint8Array(Module.Kremlin.mem.buffer)
+      .subarray(ptr, ptr + len)));
+    return new Uint8Array(result);
+  };
+
+  const callWithProto = function(proto, args, loc_name) {
+    var expected_args_number = proto.args.filter(arg => arg.interface_index !== undefined).length;
+    if (args.length != expected_args_number) {
+      throw Error("wrong number of arguments to call the F*-wasm function " + loc_name + ": expected " + expected_args_number + ", got " + args.length)
+    }
+    let memory = new Uint32Array(Module.Kremlin.mem.buffer);
+    let sp = memory[0];
+    var var_lengths = {};
+    // Populating the variable length arguments by retrieving buffer lengths
+    proto.args.map((arg) => {
+      if (arg.type === "buffer") {
+        if ((typeof arg.size === "string") && (arg.interface_index !== undefined)) {
+          let func_arg = args[arg.interface_index];
+          var_lengths[arg.size] = func_arg.length;
+        }
+      }
+      if ((arg.type === "int") && (arg.interface_index !== undefined)) {
+        let func_arg = args[arg.interface_index];
+        var_lengths[arg.name] = func_arg;
+      }
+    });
+    // Retrieving all input buffers and allocating them in the Wasm memory
+    let args_pointers = proto.args.map((arg, i) => {
+      if (arg.type === "buffer") {
+        var size;
+        if (typeof arg.size === "string") {
+          size = var_lengths[arg.size]
+        } else {
+          size = arg.size;
+        }
+        var argByteBuffer;
+        if (arg.kind === "input") {
+          let func_arg = args[arg.interface_index];
```

Comment:
what was your conclusion regarding whether the `index` was really necessary? did you end up having non-sequential numbering anywhere?

Reply:
After exposing all the functions, my conclusion is the following : what we need for sure is a `is_visible_in_exposed_js_api` boolean flag. For the order in which they appear in the interface, it happens that for all the things I exposed it follows the order of the Wasm arguments, but this is a matter of aesthetics. For instance for the Blake2b/s interface, the `output length` argument is currently first in the JS API, but you could as well put it last. The `interface_index` number gives us a little more flexibility in designing the API at the expense of numbering a few arguments explicitely, I don't think it's a horrible tradeoff so no strong feelings for either keeping it or removing it.